### DISCUSSION TITLE
Best way of using firebase with jotai?

### DISCUSSION THREAD

**Discussion by christucket** [pmndrs/jotai#754] [Q&A]
I'm trying to convert my current react useState project to use jotai to keep my local state in sync with a firebase realtime database. Some difficulties/troubles revolve around the fact that the database document (it's based on date in the url) will not be known immediately and could even be swapped mid app usage, someone can choose to work on a different date. 

I'd like to explain my current process and then maybe adapt it to jotai. First I'd need to create the "default" shape of the object if it doesn't already exist on first render. Something like this is what I have now

```javascript
// make sure the date has the initial data on the first mount
useEffect(() => {
    firebase.ref(date /* my unknown at start key */).once('value', async snapshot => {
        if (!snapshot.val()) { // that day does not currently exist
            // create a default day to start work
            await firebase.ref(date).set({
                ...defaults
            })
        } else if (!snapshot.val().player) { // basically multiple things could happen, add a new player to the game
            await database.ref(`${date}/players/${id}`).update(player)
        }
    })
}, [])
```

Once the initial database store is setup, I listen to my database like so:

```javascript
useEffect(() => {
    // anytime anything changes in the database, fetch everything and change the entire store
    database.ref(date).on('value', (snapshot) => {
        setLocalStoreState({
            ...snapshot.val(),
            ...some other stuff?
        })
    })
}, []) 
```

Then for any state updates, I have an update function to handle the object.

```javascript
const updateCellInBoard = (cell) => {
    // update one cell only
    firebase.ref(`${date}/board/${cell.id}`).update(cell)
}
const updateClock = (isPaused) => {
    firebase.ref(`${date}/appState/clock`).update(isPaused)
}
```

The (sort of) tricky part also comes from the fact that this is a typing game, so I have a global keyboard listener on the app. There would be individual grid of cells that I could type into but a global listener that keeps track of the player's cursor and updates the cell based on the cursor position and not some individual, cell-specific event handler.

My initial thoughts are: splitAtom sounds great for the board, each individual grid cell can be an atom. I can create an update atom that takes the cell and updates the board based on the id, similar to what I did before.

There are a few problems I can't wrap my head around and if you could (maybe just a fresh perspective could help) give me the recommended jotai approach to something like this, that'd help a bunch. I've seen the `onMount` function but I'm not sure how I'd use it in this case since I don't know the date immediately. In my attempts, I felt like I started writing a bunch of atoms (with a lot of update logic) that all resided outside of React, it just kind of felt "wrong".

So things I'm looking for:
1) be able to store a medium sized object of objects. Main property being an array of grid cells.
2) i'd need to be able to update each of those cells individually and without declaring any sort of useAtom() on a specific atom because there's a global keyboard handler.
3) i'd like to get away from updating the entire state on every database change and only update the specific value that gets updated. I think this just requires more dedicated listeners but I can't figure out a good paradigm using jotai. 
4) be able to mount firebase directly to my atoms and not cause some kind of feedback loop where an atom updates the database and then the database updates the atom, causing the atom to update the database again... etc
5) do it in a "jotai" way

and to answer my own questions in my basic understanding:
1) that's easy enough, just `atom({})` to start with. I could then even update the atom in the same useEffect/naive approach as before for the initial values
2) i'm not sure if this is right but I created a new write atom that could dynamically update any of the cells
```javascript
const storeAtom = atom({})
const boardAtom = atom(get => get(storeAtom).board)
const cellsAtom = splitAtom(boardAtom) // i'm not 100% sure this whole block of code works, i did something similar that kind of worked though
const updateBoardCellAtom = atom(
    null,
    (get, set, update) => {
        const cells = get(cellsAtom)
        set(cells[update.id], update)
        firebase.ref(`${date}/board/${update.id}`).update(update)  // i don't actually have a `date` here, i could store it in another atom?
    }
)
```
3) i'm not quite sure about this one. I guess I could do a lot of `useEffect()`s and update them like this, i'm looking for each atom to kind of handle this itself though
```javascript
const [attr1, setAttr1] = useAtom(attr1Atom)
const [attr2, setAttr2] = useAtom(attr2Atom)
/// ...
/// ...

useEffect(() => {
    database.ref(`${date}/attr1`).on('value', (snapshot) => {
        setAttr1(snapshot.val())
    })
    database.ref(`${date}/attr2`).on('value', (snapshot) => {
        setAttr2(snapshot.val())
    })
    // ...
})
```
i had an issue however that one update caused multiple updates to the database/state in some weird loop. 
4) no idea, i just feel like i'm doing everything above incorrectly. i've been storing/creating all of my atoms outside of any react component and in a global `store.js` file, but maybe I need to create the initial atom and do the onMount later on? I'd then maybe need to store the atom in a ref or something?
5) i think I can piecemeal everything together in a very bad way but honestly i think your work is great and i'd like to hear what the "correct" way to sync a database with my use case here. my biggest gripe with my approach is that is doesn't seem to utilize jotai functions or methodology very well and i'm looking to learn the jotai approach

Any help/feedback would be appreciated. Sorry if this isn't supposed to be a support forum but right now certain aspects of jotai examples are missing from the internet and I'm a bit lost

**Comment by dai-shi** [comment]
Let me drop some examples in general and hope you can get some ideas.

### useEffect with useAtom is fairly good

I would still say it's one of jotai practices.

If you have useState+useEffect code like this:
```jsx
const Component = () => {
  const [data, setData] = useState({})
  useEffect(() => {
    const callback = (nextData) => setData(nextData)
    const unsubscribe = subscribe(store, callback)
    return unsubscribe
  }, [])
  return <>...</>
}
```

You can simply replace useState with useAtom:
```jsx
const dataAtom = atom({})
const Component = () => {
  const [data, setData] = useAtom(dataAtom)
  useEffect(() => {
    const callback = (nextData) => setData(nextData)
    const unsubscribe = subscribe(store, callback)
    return unsubscribe
  }, [])
  return <>...</>
}
```

### onMount replaces useEffect(..., [])

Compared to useEffect, `onMount` is not easy, but it's more jotai way, because we can define them without React.

```js
const dataAtom = atom({})
dataAtom.onMount = (setAtom) => {
  const callback = (nextData) => setData(nextData)
  const unsubscribe = subscribe(store, callback)
  return unsubscribe
}
```

That's very simplistic. In reality, we would need another atom.

```js
const dataAtom = atom({})
const initDataAtom = atom(
  null,
  (get, set, init) => ...
)
initDataAtom.onMount = (initData) => {
  ...
}
```
But, it's a bit tricky, because we need to "use" initDataAtom to fire onMount.

### What if we need an input to init atom

Like your `date`. We could create an atom for it.
```js
const dataAtom = atom({})
const inputAtom = atom(null)
const initDataAtom = atom(
  null,
  (get, set, init) => {
    const input = get(inputAtom)
    ...
  }
)
initDataAtom.onMount = (initData) => {
  ...
}
```
But, this requires we need to make sure `initAtom` must be initialized before using `initDataAtom`.
If we carefully use `useEffect`, timing control might be possible.
If the `input` that is your `date` is static and can be retrieved at the root of the tree, `<Provider initialValue={[[inputAtom, '...']]}>` should work better.

So, subscription in useEffect is easier if it depends on component props.

### splitAtom usage

If you need to update an atom of split atoms, you need to pass a writable atom to `splitAtom`.
```js
const boardAtom = atom(
  get => get(storeAtom).board,
  (get, set, nextBoard) => set(storeAtom, prev => ({ ...prev, board: nextBoard }))
})
const cellsAtom = splitAtom(boardAtom)
```
Or, `focusAtom` helps for complex scenarios.

---
I don't think this this fully answers your questions.
The difficulty comes from depending on props. We would either give up with `useEffect` (delayed), use Provider if it's from the root, or redesign to have `date` in an atom from the beginning. There are some other techniques, but we'd want something fairly simple.

**Reply by christucket** [reply]
I really appreciate the response, I don't think you necessarily answered my question about firebase but you did make me feel like I wasn't too far off my approach. I'm probably over complicating it and I should just treat jotai closer to regular `useState` with some nice perks.

I wouldn't mind storing the initial value/`date` in an atom, but I think needing the data to change when `date` changes makes more sense in a useEffect anyways. Hopefully I can find a good way to write to the database without too many complications.

**Reply by dai-shi** [reply]
> I think needing the data to change when date changes makes more sense in a useEffect anyways.

Oh, I see, you want to re-subscribe when `date` changes.
At this point, I'd recommend useEffect for it. (Or, you can re-subscribe when you change `date`.)

We too would like to have a pure jotai way for re-subscription and for example `jotai/query` does that.
It's working, but it's complicated (mainly because of async/suspense). So, it's not like best practice to recommend widely yet. We would like to find an easier way or a util to hide complications.
